bpf/bpf_netdev.c

```diff
@@ -502,31 +502,55 @@ static __always_inline int do_netdev_encrypt_pools(struct __sk_buff *skb)
 	return ret;
 }
 
-static __always_inline int do_netdev_encrypt_fib(struct __sk_buff *skb, int *encrypt_iface)
+#define bpf_printk(fmt, ...)				\
+({							\
+	char ____fmt[] = fmt;				\
+	trace_printk(____fmt, sizeof(____fmt),	\
+			 ##__VA_ARGS__);		\
+})
+
+static __always_inline int do_netdev_encrypt_fib(struct __sk_buff *skb,
+						 __u16 proto,
+						 int *encrypt_iface)
 {
 	int ret = 0;
 
 #ifdef HAVE_FIB_LOOKUP
 	struct bpf_fib_lookup fib_params = {};
 	void *data, *data_end;
-	struct iphdr *iphdr;
 	__be32 sum;
 	int err;
 
-	if (!revalidate_data(skb, &data, &data_end, &iphdr)) {
-		ret = DROP_INVALID;
-		goto drop_err_fib;
+	if (proto ==  bpf_htons(ETH_P_IP)) {
+		struct iphdr *ip4;
+
+		if (!revalidate_data(skb, &data, &data_end, &ip4)) {
+			ret = DROP_INVALID;
+			goto drop_err_fib;
+		}
+
+		fib_params.family = AF_INET;
+		fib_params.ipv4_src = ip4->saddr;
+		fib_params.ipv4_dst = ip4->daddr;
+	} else {
+		struct ipv6hdr *ip6;
+
+		if (!revalidate_data(skb, &data, &data_end, &ip6)) {
+			ret = DROP_INVALID;
+			goto drop_err_fib;
+		}
+
+		fib_params.family = AF_INET6;
+		ipv6_addr_copy((union v6addr *) &fib_params.ipv6_src, (union v6addr *) &ip6->saddr);
+		ipv6_addr_copy((union v6addr *) &fib_params.ipv6_dst, (union v6addr *) &ip6->daddr);
 	}
 
-	fib_params.family = AF_INET;
 	fib_params.ifindex = *encrypt_iface;
 
-	fib_params.ipv4_src = iphdr->saddr;
-	fib_params.ipv4_dst = iphdr->daddr;
-
 	err = fib_lookup(skb, &fib_params, sizeof(fib_params),
 		    BPF_FIB_LOOKUP_DIRECT | BPF_FIB_LOOKUP_OUTPUT);
 	if (err != 0) {
+		bpf_printk("fib lookup failed proto %u ... dev %d err %d\n", proto, *encrypt_iface, err);
```

Comment:
We should really forward this information up through the monitor rather than spamming the trace_pipe.

Reply:
FWIW I have another series I'll post monday to push the fib lookup error code through monitor so we can see what the actual error is. The FIB_* errors hide what the actual error in datapath was.